### PR TITLE
feat: use yq to parse yaml file

### PR DESCRIPTION
**Pull request by gregmagolan**
- Use yq to parse yaml file
- wip

---

### Type of change

- Bug fix (change which fixes an issue)
- New feature or functionality (change which adds functionality)
- Style (white-space, formatting, etc...)
- Refactor (a code change that neither fixes a bug or adds a new feature)
- Performance (a code change that improves performance)
- Documentation (updates to documentation or READMEs)
- Chore (any other change that doesn't affect source or test files, such as configuration)

**For changes visible to end-users**

- Breaking change (this change will force users to change their own code or config)
- Relevant documentation has been updated
- Suggested release notes are provided below:

### Test plan

- Covered by existing test cases
- New test cases added
- Manual testing; please provide instructions so we can reproduce:

### CODE REVIEW

File: npm/private/npm_translate_lock.bzl [hunk 1]
```diff
@@ -70,13 +73,12 @@ _ATTRS = {
     "repositories_bzl_filename": attr.string(default = DEFAULT_REPOSITORIES_BZL_FILENAME),
     "root_package": attr.string(default = DEFAULT_ROOT_PACKAGE),
     "update_pnpm_lock": attr.bool(),
-    "update_pnpm_lock_node_toolchain_prefix": attr.string(),
     "use_home_npmrc": attr.bool(),
+    "use_starlark_yaml_parser": attr.bool(),
```
Comment (jbedard) [inline npm/private/npm_translate_lock.bzl:77]: Is there a reason we have to keep it? Can we mark it as deprecated to delete in 2.0?

File: npm/private/npm_translate_lock.bzl [hunk 2]
```diff
@@ -485,14 +494,42 @@ def npm_translate_lock(
...
+
+            The `?` character in the `tarball` value causes `yq` to fail with:
+
+            ```
+            $ yq pnpm-lock.yaml -o=json
+            Error: bad file 'pnpm-lock.yaml': yaml: line 7129: did not find expected ',' or '}'
+            ```
+
+            If the tarball value is quoted then yq would accept it but as of this writing, the latest
+            version of pnpm (8.14.3) does not quote such a value and the latest version of yq (4.40.5)
```
Comment (jbedard) [inline npm/private/npm_translate_lock.bzl:517]: Is there a `yq` or `pnpm` bug we can link to?

File: npm/private/test/package.json [hunk 3]
```diff
@@ -2,7 +2,7 @@
     "private": true,
     "// devDependencies": "test linking & building 3rd party npm packages with problematic postinstall",
     "devDependencies": {
-        "@blockprotocol/type-system-web": "https://gitpkg.vercel.app/blockprotocol/blockprotocol/packages/%40blockprotocol/type-system-web?6526c0e",
+        "hello": "https://gitpkg.vercel.app/EqualMa/gitpkg-hello/packages/hello",
```
Comment (jbedard) [inline npm/private/test/package.json:5]: What is this?

File: npm/private/test/parse_pnpm_lock_tests.bzl [hunk 4]
```diff
@@ -6,15 +6,15 @@ load("//npm/private:utils.bzl", "utils")
 def _parse_empty_lock_test_impl(ctx):
     env = unittest.begin(ctx)
 
-    parsed = utils.parse_pnpm_lock("")
+    parsed = utils.parse_pnpm_lock_yaml("")
```
Comment (jbedard) [inline npm/private/test/parse_pnpm_lock_tests.bzl:9]: I think we need to test both `parse_pnpm_lock_*` properly and not pick&choose like this. Especially after https://github.com/aspect-build/rules_js/pull/1454

File: npm/private/utils.bzl [hunk 5]
```diff
@@ -146,8 +146,28 @@ def _parse_pnpm_lock(content):
...
         A tuple of (importers dict, packages dict)
     """
-    parsed = _parse_yaml(content)
+    return _parse_pnpm_lock_common(_parse_yaml(content))
 
+def _parse_pnpm_lock_json(content):
+    """Parse the content of a pnpm-lock.yaml file.
+
+    Args:
+        content: lockfile content
```
Comment (jbedard) [inline npm/private/utils.bzl:155]: * as json

File: npm/private/test/parse_pnpm_lock_tests.bzl (hunk 4 above)
Comment (jbedard) [reply npm/private/test/parse_pnpm_lock_tests.bzl:9]: Actually now (after scrolling more) I see they share the main implementation so most of the testing doesn't need to be done with both? Except parsing edge cases such as this one (empty files) and things like invalid yaml in https://github.com/aspect-build/rules_js/pull/1454

File: npm/private/npm_translate_lock.bzl (hunk 1 above)
Comment (gregmagolan) [reply npm/private/npm_translate_lock.bzl:77]: See PR summary and docstring for this attribute for reasoning. We could remove this in the future once the fix for https://github.com/pnpm/pnpm/issues/5414 lands and settles but there is no rush since its a simple logic branch to maintain.

File: npm/private/npm_translate_lock.bzl (hunk 2 above)
Comment (gregmagolan) [reply npm/private/npm_translate_lock.bzl:517]: Ahh. Thanks for the find @DavidZbarsky-at

File: npm/private/test/package.json (hunk 3 above)
Comment (gregmagolan) [reply npm/private/test/package.json:5]: Related to https://github.com/pnpm/pnpm/issues/5414. I'll pre-factor this and self-contained changes out.

File: npm/extensions.bzl [hunk 6]
```diff
@@ -63,7 +65,23 @@ def _extension_impl(module_ctx):
             if not attr.pnpm_lock:
                 continue
 
-            lock_importers, lock_packages, lock_patched_dependencies = utils.parse_pnpm_lock(module_ctx.read(attr.pnpm_lock))
+            if attr.use_starlark_yaml_parser:
```
Comment (alexeagle) [inline npm/extensions.bzl:68]: maybe this should WARN since we'd like to drop the yaml parser code eventually?

File: npm/private/npm_import.bzl [hunk 7]
```diff
@@ -420,7 +420,12 @@ def _is_gnu_tar(rctx):
     if repo_utils.is_linux(rctx):
         return True
 
-    result = rctx.execute(["tar", "--version"])
+    tar_args = ["tar", "--version"]
```
Comment (alexeagle) [inline npm/private/npm_import.bzl:423]: TODO to use hermetic tar toolchain?

File: npm/private/npm_translate_lock_state.bzl [hunk 8]
```diff
@@ -140,6 +140,9 @@ def _init_common_labels(priv, rctx, label_store):
     # pnpm-workspace.yaml file
     label_store.add_sibling("lock", "pnpm_workspace", PNPM_WORKSPACE_FILENAME)
 
+    # yq is used for parsing the pnpm lock file (WORKSPACE only)
```
Comment (alexeagle) [inline npm/private/npm_translate_lock_state.bzl:143]: why WORKSPACE only? That seems scary

Comment (gregmagolan) [reply npm/private/npm_translate_lock_state.bzl:143]: actually, that comment is out-dated. I figured out how to wire it up to in the module extension context as well now.

File: npm/private/npm_import.bzl (hunk 7 above)
Comment (gregmagolan) [reply npm/private/npm_import.bzl:423]: In theory we could, but tar happens to ship by default on macos, every major linux distro and even on Windows now. But I'll drop a TODO there.

File: npm/private/test/package.json (hunk 3 above)
Comment (jbedard) [reply npm/private/test/package.json:5]: Thanks for pulling those out 👍

File: npm/private/npm_import.bzl (hunk 7 above)
Comment (gregmagolan) [reply npm/private/npm_import.bzl:423]: Benefit of using a hermetic tar would be that we don't have to check the version and can remote the "if gnu-tar" branch that adds an additional flag.

Comment (gregmagolan) [reply npm/private/npm_import.bzl:423]: Addressed in https://github.com/aspect-build/rules_js/pull/1460

File: npm/extensions.bzl (hunk 6 above)
Comment (gregmagolan) [reply npm/extensions.bzl:68]: given that some users will be force into the yaml parser due to https://github.com/pnpm/pnpm/issues/5414, I'd rather wait on the warning until that issue is resolved and the warning can suggest to upgrade to the latest pnpm version.

File: npm/private/utils.bzl (hunk 5 above)
Comment (gregmagolan) [reply npm/private/utils.bzl:155]: 👍

**Comment by jbedard** [comment]
Can we land https://github.com/aspect-build/rules_js/pull/1454 first and ensure we keep the same behaviour with the new parser?

File: npm/private/npm_translate_lock.bzl (hunk 2 above)
Comment (gregmagolan) [reply npm/private/npm_translate_lock.bzl:517]: Added a link to this issue to the docstring here.

**Review by jbedard — APPROVED** [review]
👍
